ledger-api/grpc-definitions/com/digitalasset/admin/api/v1/package_management_service.proto

```diff
@@ -0,0 +1,100 @@
+service PackageManagementService {
+
+  // Returns the details of all supported packages.
+  rpc ListPackageDetails (ListPackageDetailsRequest) returns (ListPackageDetailsResponse);
+
+  // Uploads a DAR archive.
+  rpc UploadArchive (UploadArchiveRequest) returns (UploadArchiveResponse);
+
+}
+
+message ListPackageDetailsRequest {
+
+  // Must correspond to the ledger ID reported by the Ledger Identification Service.
+  // Required
+  string ledger_id = 1;
+}
+
+message ListPackageDetailsResponse {
+
+  // The IDs of all DAML-LF packages supported by the server.
+  // Required
+  repeated string PackageDetails = 1;
+}
+
+message PackageDetails {
+ 
+  // The IDs of all DAML-LF packages supported by the server.
+  // Required
+  string package_id = 1;
+
+  // Human readable name of the package.
+  // Optional
+  string package_name = 2;
+
+  // Identifier of the agent that uploaded the package. Depending on the ledger 
+  // implementation, the interpretation of this field may differ. It could be
+  // the identifier of the uploading application or its user. It could also
+  // be the unique identifier of the participant node within the cluster.
+  // Optional
+  string uploaded_by = 3;
+
+  // Microseconds since the UNIX epoch. Can go backwards. Fixed
+  // since the vast majority of values will be greater than
+  // 2^28, since currently the number of microseconds since the
+  // epoch is greater than that. Range: 0001-01-01T00:00:00Z to
+  // 9999-12-31T23:59:59.999999Z, so that we can convert to/from
+  // https://www.ietf.org/rfc/rfc3339.txt.
+  // Required
+  google.protobuf.Timestamp upload_timestamp = 4;
+
+  // Size of the package in bytes.
+  // Required
+  uint64 package_size = 5 [jstype = JS_STRING];
+
+  // names of the modules contained in the package
+  // Optional
+  repeated string module_names = 6;
+}
+
+message UploadDarRequest {
+
+  // Must correspond to the ledger ID reported by the Ledger Identification Service.
+  // Required
+  string ledger_id = 1;
+
+  // Contains a DAML archive DAR file, which in turn is a jar like zipped container
+  // for ``daml_lf`` archives. See further details in ``daml_lf.proto``.
+  // Required
+  bytes dar = 2;
+
+  // Name of the DAR file. Implementation is ikely to store it in a local repository.
+  // This name will be likely used as part of the name in the local file system.
+  // Please note that on the query side, we no longer deal with DARs, but we deal with
+  // packages instead. There is a 1 to N relationship between DARs and packages.
+  // Required
+  string filename = 3;
+
+  // Identifier of the agent that uploaded the package. Depending on the ledger 
+  // implementation, the interpretation of this field may differ. It could be
+  // the identifier of the uploading application or its user. It could also
+  // be the unique identifier of the participant node within the cluster.
+  // Optional
+  string uploaded_by = 4;
```

Comment:
If this is supposed to be metadata that is user-supplied, might be better to simply make this a more general key-value mapping than naming a specific field "uploaded_by" and not allowing room for other metadata (such as file modified, hash of the DAR), or simply excluding this altogether and let implementations define how they manage packages. This and the "filename" field feel somewhat out of place to me.

Reply:
I don't really think we should allow any user-supplied metadata. The package uploads are ledger-wide and many packages are common across everything (standard library for example). I wouldn't store anything else in the ledger except timestamp (record time of the submission), the submitting party and the daml-lf archive bytes.

Reply:
The submitting party does not exist on admin API. Remember, this is not a Ledger API which is party-level, but Admin API, which is participant level. In the discussions I have had with various stakeholders on this subject, the recurring theme was to have an indication of who uploaded the package. It could be that we just stay with the participant identifier instead of this field. We need to figure out how do we name the participants in the cluster. It seems like so far name assignment has been implementation specific. We might want to change that though by introducing another call.